File: src/api/logMessage.ts
export enum LogLevel {
  Debug = 0,
  Info,
  Warning,
  Error,
  Critical,
}

export interface LogMessage {
  level: LogLevel;
  message: string;
  sourcePath?: string | null;
  sourceLine?: number | null;
  time: string; // ISO8601 string
  tags?: string[] | null;
}

File: backend/Logs/LogMessage.cs
using System.Text.Json.Serialization;
using Olve.Trains.UI.Server.Api;

namespace Olve.Trains.UI.Server.Logs;

public sealed record LogMessage(
    LogLevel Level,
    string Message,
    [property: JsonConverter(typeof(PathJsonConverter))]
    IPath? SourcePath,
    int? SourceLine,
    DateTime Time,
    string[]? Tags
);